src/py/flwr/client/typing.py

```diff
@@ -15,14 +15,14 @@
 """Custom types for Flower clients."""
 
 
-from typing import Callable
+from typing import Callable, Optional
 
 from flwr.common import Context, Message
 
 from .client import Client as Client
 
 # Compatibility
-ClientFn = Callable[[str], Client]
```

Comment:
Should we keep this as `ClientFnLegacy` or so?

This type is still used in `start_simulation` and `start_client`, or not?

Reply:
we can keep the previous `ClientFn` as `ClientFnLegacy`. 

Not sure if I understand the second question, both go through the same same logic in `message_handler` so they will both fetch the `partition_id` from the `Context` and `node_id` from the `Message`. However, since `start_client` isn't triggered via `flower-supernode`, `partition_id` will remain unset. As for `node_id`, it defaults to `0` for `start_client` (probably we can changes things in our compatibility layer so it uses a randomly generate value). in case of `start_simulation` it will use an UUID once #3699 is merged.

Reply:
Let me rephrase. Existing projects that use either `start_simulation` or `start_client` might depend on `ClientFn` if they use type declarations. After more digging I noticed that it's exported and part of the public API. I'm therefore retracting my comment above - we must keep it as-is and introduce the new signature under a new name (`ClientFn...`).

Reply:
The `message_handler` question is more difficult. It seems like there is no issue with everything related to Flower Next. It would be nice to find a better solution for `start_client` (something other than passing 0), but it's also not too critical. `node_id 0` and `partition_id None` might be the best solution - it'd be confusing to see a different random `node_id` on the client and on the server.